bin/test_configure_tower.py

```diff
@@ -0,0 +1,237 @@
+        os.walk.assert_called_once()
+
+    def test_load_projects_valid(self, projects, mocker):
+        return_value = iter(self.valid_project_filenames)
+        mocker.patch.object(projects, "list_projects", return_value=return_value)
+        mocker.patch("builtins.open")
+        mocker.patch("yaml.load", side_effect=iter(self.valid_projects))
+        expected = self.valid_projects
+        actual = list(projects.load_projects())
+        assert actual == expected
+        projects.list_projects.assert_called_once()
+        open.assert_called_with(self.valid_project_filenames[-1])
+
+    def test_load_projects_invalid(self, projects, mocker):
+        return_value = iter(self.valid_project_filenames)
+        mocker.patch.object(projects, "list_projects", return_value=return_value)
+        mocker.patch("builtins.open")
+        invalid_projects = [{"foo": "bar"}, {"template_path": "baz"}]
+        mocker.patch("yaml.load", side_effect=iter(invalid_projects))
+        expected = list()
+        actual = list(projects.load_projects())
+        assert actual == expected
+        projects.list_projects.assert_called_once()
+        open.assert_called_with(self.valid_project_filenames[-1])
+
+    def test_extract_emails_valid(self, projects):
+        input = self.valid_arns
+        expected = self.valid_emails
+        actual = projects.extract_emails(input)
+        assert actual == expected
+
+    def test_extract_emails_empty(self, projects):
+        input = ""
+        expected = []
+        actual = projects.extract_emails(input)
+        assert actual == expected
+
+    def test_extract_emails_invalid(self, projects):
+        for arn in self.invalid_arns:
+            with pytest.raises(ValueError):
+                projects.extract_emails([arn])
+
+    def test_extract_users_valid(self, mocker):
+        mocker.patch.object(
+            configure_tower.Projects,
+            "load_projects",
+            return_value=iter(self.valid_projects),
+        )
+        mocker.patch.object(
+            configure_tower.Projects,
+            "extract_emails",
+            side_effect=lambda x: ["foo"] if len(x) > 0 else [],
+        )
+        mocker.patch.object(configure_tower, "Users")
+        actual = configure_tower.Projects(CONFIG_DIR)
+        assert list(actual.users_per_project.keys()) == self.valid_project_names
+        actual.load_projects.assert_called_once()
+        configure_tower.Users.assert_has_calls(
+            [
+                mocker.call(maintainers=["foo"], viewers=[]),
+                mocker.call(maintainers=[], viewers=["foo"]),
+            ]
+        )
+
+    def test_extract_users_invalid(self, mocker):
+        for invalid_project in self.invalid_projects:
+            mocker.patch.object(
+                configure_tower.Projects,
+                "load_projects",
+                return_value=iter([invalid_project]),
+            )
+            with pytest.raises(AssertionError):
```

Comment:
I know you're not done with the PR, but there are missing invalid test cases here.

Reply:
Which ones should be added? How many is appropriate to test for?

Reply:
Ideally, you test for all three assertions, since there are three different cases that an assertion error is thrown.  That being said this could be overkill here.

One definitely over the top thing you can do is create a function for the validation:

```
def _validate_config():
     # assertions here
```

You then test that the three asserts happen for `_validate_config` and when you're testing the `load_projects` function, all you have to test for is that you've called `_validate_config` at least once.  That also could reduce the amount of class attributes you have in your Test class.  Again, I don't think it's necessary here, just some food for thought.

Reply:
I like the idea of a `validate_config()` function, especially since it helps simplify the target code and the tests. I implemented it in #64 (ea3079ad6206c48941d5f550fd0c8b78ba0783c6) and updated the tests in this PR (29541fcc0c42f5748e787a4a3caefaae49469b93).